rskj-core/src/main/java/org/ethereum/rpc/LogFilter.java

```diff
@@ -231,47 +232,75 @@ private static void processSingleBlockByHash(String blockHash, Blockchain blockc
-        BlocksBloom auxiliaryBlocksBloom = null;
-        long bestBlockNumber = blockchain.getBestBlock().getNumber();
+    private static void processBlocks(Block fromBlock, Block toBlock, LogFilter filter, Blockchain blockchain, BlocksBloomStore blocksBloomStore) {
+        final long bestBlockNumber = blockchain.getBestBlock().getNumber();
 
-        for (long blockNum = fromBlockNumber; blockNum <= toBlockNumber; blockNum++) {
-            boolean isConfirmedBlock = blockNum <= bestBlockNumber - blocksBloomStore.getNoConfirmations();
+        BlocksBloom bloomAccumulator = null;
 
-            if (isConfirmedBlock) {
-                if (blocksBloomStore.firstNumberInRange(blockNum) == blockNum) {
-                    if (blocksBloomStore.hasBlockNumber(blockNum)) {
-                        BlocksBloom blocksBloom = blocksBloomStore.getBlocksBloomByNumber(blockNum);
+        Block block = toBlock;
+        while (block != null && block.getNumber() >= fromBlock.getNumber()) {
+            boolean isConfirmedBlock = block.getNumber() <= bestBlockNumber - blocksBloomStore.getNoConfirmations();
 
-                        if (!filter.addressesTopicsFilter.matchBloom(blocksBloom.getBloom())) {
-                            blockNum = blocksBloomStore.lastNumberInRange(blockNum);
-                            continue;
-                        }
-                    }
+            BlocksBloom blocksBloom = isConfirmedBlock ? blocksBloomStore.getBlocksBloomByNumber(block.getNumber()) : null;
+            boolean hasBloom = blocksBloom != null;
 
-                    auxiliaryBlocksBloom = new BlocksBloom();
-                }
+            if (hasBloom && !bloomMatches(blocksBloom, filter)) {
+                block = getBlockToSkipToOrNullIfDone(block, fromBlock.getNumber(), blocksBloomStore, blockchain);
+                continue;
+            }
 
-                Block block = blockchain.getBlockByNumber(blockNum);
+            boolean needsBloom = isConfirmedBlock && !hasBloom;
+            if (needsBloom) {
+                bloomAccumulator = addBlockBloom(block, bloomAccumulator, blocksBloomStore);
+            }
 
-                if (auxiliaryBlocksBloom != null) {
-                    auxiliaryBlocksBloom.addBlockBloom(blockNum, new Bloom(block.getLogBloom()));
-                }
+            // reverseTxOrder=true to process txs in reverse order for later complete list reverse
+            filter.onBlock(block, true);
+            block = blockchain.getBlockByHash(block.getParentHash().getBytes());
+        }
 
-                if (auxiliaryBlocksBloom != null && blocksBloomStore.lastNumberInRange(blockNum) == blockNum) {
-                    blocksBloomStore.addBlocksBloom(auxiliaryBlocksBloom);
-                }
+        // sort in a from-to fashion after looping in reverse order
+        filter.reverseEvents();

```

Comment:
maybe the filter already have events, so maybe it could be better to only revert these events and not all the filter list.

Reply:
It doesn't, Filter was created specifically for this "backwards" round.